### DISCUSSION TITLE
Do you still have a class library for developing panoramic functions in AS3?

### DISCUSSION THREAD

**Discussion by Jason0247** [airsdk/Adobe-Runtime-Support#2907] [Q&A] [answered]
Recently, there is a project that requires the development of panoramic functions, which require the use of hexahedra and the ability to dynamically add hotspots and other functions.
In the early years, there were many such libraries, but now we can no longer find resources. Do you still have relevant open source libraries?
I originally wanted to rewrite it using away3D, but the information is really difficult to find...

**Comment by xabip** [comment]
Hi Jason,
 
I don't think there's any specific library out there right now... We used krPano, but we had to create our library to read their format and generate the panoramas (with hotspots and navigation) with Starling and their (small) 3d capabilities. I can't share it, unfortunately. 
Alternatively, I suggest using a webview on top of the display list where you can use javascript if that's possible for your use case.

I hope this helps.

**Reply by Jason0247** [reply]
HI, xabip
First of all, I would like to express my gratitude. In fact, I have also tried using the local package of krPano, but directly reading Html will result in an error. I heard that it seems that cross domain access is necessary, but unfortunately, I did not find similar properties or methods in AIR, so I cannot use panoramic packages like krPano locally.
When I tried to upload krPano to the server and access it remotely, it was able to display it, but it got stuck. I don't know why, because I couldn't use local packages to display it, I gave up using krPano.

**Reply by Jason0247** [reply]
Error #2044: Unhandled ErrorEvent:. text=The protocol is not supported.
When I try to access the local krpano, I will be prompted for this.

**Reply by ajwfrost** [reply]
Unsupported protocol implies you're not using http/https to access it...? if this is something that needs to be run on a webserver, it's probably simpler to have that set up remotely (you may need to look at its configuration if you're doing cross-origin requests maybe?). If you wanted to keep it running locally, then there's a solution I know about that may help.. but I don't really know your use case here I'm afraid, not sure how much of an ActionScript application you need to create or whether you just need to host this existing HTML/JS based component?

thanks

**Reply by xabip** [reply, marked as answer]
To overcome this issue, we had a small server running locally (internally in the AIR app) serving content over the network in a specific port. This way, it's all contained within the app and you control what you load (including offline functionality). There's a security message enabled in the browsers that prevents loading files directly from the filesystem in JS. We would be loading both the HTML and all it's content, including KRpano libraries and assets from this local server so everything works in the same origin. (Meeting the Same-Origin Policy restrictions)

**Reply by Jason0247** [reply]
@ajwfrost 
Yes, it is easy to load on the server (but the operation feels too laggy, I don't know why)
However, most of the time, it is necessary to deploy and load the HTML file of krPano locally, which leads to cross domain access issues.
<img width="1244" alt="ddc76a21bca6ec291ce027ca12de719" src="https://github.com/airsdk/Adobe-Runtime-Support/assets/91530077/7ca21e29-1a0b-415e-bec1-5378678035a6">

When developing native iOS, adding this code will allow for cross domain, but I don't know how AIR needs to handle it
![f670bfbbf81bf91ceb13726376a918e](https://github.com/airsdk/Adobe-Runtime-Support/assets/91530077/cd0eb50d-2cb2-4413-9be7-2b29816e186d)
thanks

**Reply by Jason0247** [reply]
@xabip I roughly understand what you mean. In fact, I created a small server through AIR, and when loading HTML, I used a local area network address instead of local HTML access.
I have considered it before, but there were various issues that prevented it from loading properly, so I had to give up.

**Comment by Jason0247** [comment]
Hello, excuse me, but I still have a question to ask.
Can you display panoramic views normally on the Android system through stageWebView?
I also used KrPano, which can be displayed normally on Windows but not on Android. The menu is normal, but the panoramic part is not displayed. Did I set it wrong somewhere?
thanks @xabip

**Comment by cairookie** [comment]
away3d清晰度很不好。。我们使用了away3d开发了全景功能。
